src/librustc_mir/transform/type_check.rs

```diff
@@ -1110,21 +1151,123 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
     }
 
     fn check_rvalue(&mut self, mir: &Mir<'tcx>, rvalue: &Rvalue<'tcx>, location: Location) {
+        let tcx = self.tcx();
+
         match rvalue {
             Rvalue::Aggregate(ak, ops) => {
                 self.check_aggregate_rvalue(mir, rvalue, ak, ops, location)
             }
+
+            Rvalue::Repeat(operand, const_usize) => if const_usize.as_u64() > 1 {
+                let operand_ty = operand.ty(mir, tcx);
+
+                let trait_ref = ty::TraitRef {
+                    def_id: tcx.lang_items().copy_trait().unwrap(),
+                    substs: tcx.mk_substs_trait(operand_ty, &[]),
+                };
+
+                self.prove_trait_ref(trait_ref, location);
+            },
+
+            Rvalue::NullaryOp(_, ty) => {
+                let trait_ref = ty::TraitRef {
+                    def_id: tcx.lang_items().sized_trait().unwrap(),
+                    substs: tcx.mk_substs_trait(ty, &[]),
+                };
+
+                self.prove_trait_ref(trait_ref, location);
+            }
+
+            Rvalue::Cast(cast_kind, op, ty) => match cast_kind {
+                CastKind::ReifyFnPointer => {
+                    let fn_sig = op.ty(mir, tcx).fn_sig(tcx);
+
+                    // The type that we see in the fcx is like
+                    // `foo::<'a, 'b>`, where `foo` is the path to a
+                    // function definition. When we extract the
+                    // signature, it comes from the `fn_sig` query,
+                    // and hence may contain unnormalized results.
+                    let fn_sig = self.normalize(&fn_sig, location);
+
+                    let ty_fn_ptr_from = tcx.mk_fn_ptr(fn_sig);
+
+                    if let Err(terr) = self.eq_types(ty_fn_ptr_from, ty, location.at_self()) {
```

Comment:
I think typeck allows some sort of (e.g. higher-ranked) subtyping here, so you'll need to allow it too.

e.g.
```Rust
// this sig contains a binder
fn foo<'a>(_x: &'a ()) {}
fn bar<'b>() {
    foo as fn(&'b ()); // ... but this doesn't
}
```

Reply:
Hmm, that code type-checks. Perhaps because I haven't addressed previous comment yet.